lizard_auth_server/forms.py

```diff
@@ -283,10 +287,80 @@ def clean(self):
             if organisation_role.role.code == BILLING_ROLE and
             organisation_role.role.portal.name == THREEDI_PORTAL])
         if num_threedi_billing_roles == 0:
-            self._errors['roles'] = self.error_class([_('required 3Di billing role is missing')])
+            self._errors['roles'] = self.error_class([
+                _('required 3Di billing role is missing')])
         if num_threedi_billing_roles >= 2:
             self._errors['roles'] = self.error_class([
                 _('Only one 3Di billing role is allowed ({} found)').format(
                     num_threedi_billing_roles)])
         print(self._errors)
         return cleaned_data
+
+
+class JWTAuthenticationForm(AuthenticationForm):
+    """An authentication form for requesting JSON Web Tokens.
+
+    This form resembles the standard authentication form. It has one extra
+    hidden `portal` field for passing the public key around of the portal
+    that the user wants access to.
+
+    """
+    portal = forms.CharField(widget=forms.HiddenInput())
+
+    def clean_portal(self):
+        """Check if the sso_key exists.
+
+        Since the `portal` field is hidden, any field errors are hidden too.
+        For that reason, if an error occurs, it is also registered as a
+        form-wide error, so it will be visible to end users (form-
+        wide errors are displayed at the top of the page).
+
+        It's unlikely that this check does not pass, because validation has
+        also been done on the GET request - unless somebody is tampering
+        with the form data...
+
+        Returns:
+            Portal: an instance of lizard_auth_server.models.Portal.
+
+        Raises:
+            ValidationError: If no portal with this sso_key exists.
+
+        """
+        data = self.cleaned_data['portal']
```

Comment:
Ik zou het niet heel algemeen "data" noemen (zeker omdat je het uit `cleaned_data` haalt) maar of `portal` of `sso_key`, dat zegt meer.